Sources/Profiler.swift

```diff
@@ -0,0 +1,326 @@
+//
+//  ProcedureKit
+//
+//  Copyright © 2016 ProcedureKit. All rights reserved.
+//
+
+
+import Foundation
+
+public enum ProcedureEvent: Int {
+    case attached = 0, started, cancelled, produced, finished
+}
+
+extension ProcedureEvent: CustomStringConvertible {
+    public var description: String {
+        switch self {
+        case .attached: return "Attached"
+        case .started: return "Started"
+        case .cancelled: return "Cancelled"
+        case .produced: return "Produced"
+        case .finished: return "Finished"
+        }
+    }
+}
+// MARK: - ProcedureProfilerReporter
+public protocol ProcedureProfilerReporter {
+    func finishedProfilingWithResult(result: ProfileResult)
+}
+
+// MARK: - ProfileResult
+public struct ProfileResult {
+    public let identity: Procedure.Identity
+    public let created: TimeInterval
+    public let attached: TimeInterval
+    public let started: TimeInterval
+    public let cancelled: TimeInterval?
+    public let finished: TimeInterval?
+    public let children: [ProfileResult]
+}
+
+extension PendingValue where T: Equatable {
+    var pending: Bool {
+        if case .pending = self {
+            return true
+        }
+        return false
+    }
+
+    static func==(lhs: PendingValue<T>, rhs: PendingValue<T>) -> Bool {
+        switch (lhs, rhs) {
+        case (.pending, .pending):
+            return true
+        case (.void, .void):
+            return true
+        case let (.ready(lhsValue), .ready(rhsValue)):
+            return lhsValue == rhsValue
+        default:
+            return false
+        }
+    }
+}
+
+// MARK: - PendingResult
+
+struct PendingResult {
+
+    let created: TimeInterval
+    let identity: PendingValue<Procedure.Identity>
+    let attached: PendingValue<TimeInterval>
+    let started: PendingValue<TimeInterval>
+    let cancelled: PendingValue<TimeInterval>
+    let finished: PendingValue<TimeInterval>
+    let children: [ProfileResult]
+
+    var pending: Bool {
+        return identity.pending || attached.pending || started.pending || (cancelled.pending && finished.pending)
```

Comment:
And I guess this should be renamed `isPending` - this is just to be more consistent with `Foundation.Operation` properties, e.g. `isFinished` etc.